consensus/src/pending_votes.rs

```diff
@@ -291,23 +291,35 @@ impl PendingVotes {
         vote: Vote,
     ) -> VoteReceptionResult {
         let li_digest = vote.ledger_info().hash();
-        let (_, li_with_sig) = self.li_digest_to_votes.get_mut(&li_digest).unwrap();
-        match validator_verifier.check_voting_power(li_with_sig.signatures().keys(), true) {
-            // a quorum of signature was reached, a new QC is formed
-            Ok(_) => Self::aggregate_qc_now(validator_verifier, li_with_sig, vote.vote_data()),
-
-            // not enough votes
-            Err(VerifyError::TooLittleVotingPower { .. }) => {
-                panic!("Delayed QC aggregation should not be triggered if we don't have enough votes to form a QC");
-            },
+        match self.li_digest_to_votes.get_mut(&li_digest) {
+            Some((_, li_with_sig)) => {
+                match validator_verifier.check_voting_power(li_with_sig.signatures().keys(), true) {
+                    // a quorum of signature was reached, a new QC is formed
+                    Ok(_) => {
+                        Self::aggregate_qc_now(validator_verifier, li_with_sig, vote.vote_data())
+                    },
+
+                    // not enough votes
+                    Err(VerifyError::TooLittleVotingPower { .. }) => {
+                        panic!("Delayed QC aggregation should not be triggered if we don't have enough votes to form a QC");
+                    },
 
-            // error
-            Err(error) => {
+                    // error
+                    Err(error) => {
+                        error!(
+                            "MUST_FIX: vote received could not be added: {}, vote: {}",
+                            error, vote
+                        );
+                        VoteReceptionResult::ErrorAddingVote(error)
+                    },
+                }
+            },
+            None => {
                 error!(
-                    "MUST_FIX: vote received could not be added: {}, vote: {}",
-                    error, vote
+                    "No LedgerInfoWithSignatures found for the given digest: {}",
+                    li_digest
                 );
-                VoteReceptionResult::ErrorAddingVote(error)
+                VoteReceptionResult::ErrorAddingVote(VerifyError::EmptySignature)
```

Comment:
this is not empty signature, I don't think it's helpful to turn this unwrap into error too. it's invariant that we already have the LI to trigger this function.

Reply:
I can revert the change and document the invariant. 
I think any further change needs some discussion and alignment on how to proceed. For example, if we document invariants, do we have an automated test to catch them if something changes? Or, do we guarantee sanity checks as early as possible only once or every time? If a check is done once, how do we guarantee all execution paths are good in the next 6 months and in one year? cc @JoshLind @gedigi

Reply:
I'm not sure if it's possible to "guarantee" those invariants are hold with new development. imo with best effort testing, fail fast is much simpler than fail safe in case of bugs. fail safe (with errors) can lead to cascading failures that are much harder to debug because of some invariants are not hold. fail fast on invariants (direct panic) is simpler to see what's wrong. but it's also case by case, some times fail safe doesn't have much consequences.

Reply:
It's a bit difficult to decide what it's an invariant violation and what not if those are not documented. It's okay for me to fail fast if those are documented as such and clear.